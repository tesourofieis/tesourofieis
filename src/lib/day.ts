import { getDate, getDay, getMonth } from "date-fns";
import {
  COMMUNE_C_10A,
  COMMUNE_C_10B,
  COMMUNE_C_10C,
  COMMUNE_C_10PASC,
  COMMUNE_C_10T,
  PATTERN_ADVENT,
  PATTERN_EASTER,
  PATTERN_EPIPHANY,
  PATTERN_LENT,
  PATTERN_POST_EPIPHANY_SUNDAY,
  PATTERN_PRE_LENTEN,
  TABLE_OF_PRECEDENCE,
  TEMPORA_RANK_MAP,
  TITLES,
  TYPE_TEMPORA,
} from "./constants.ts";

// """ Class used to keep `Observance` objects for particular days of Missal.
//
// It contains three lists: `tempora`, `celebration` and `commemoration`.
// On Missal's creation the lists are filled in so that `tempora` always contains `Observance` representing
// given variable day, `celebration` contains an `Observance`s to be celebrated in this day and
// `commemoration` contains zero or more `Observance`s that should be commemorated with the main celebration.
// """
class Day {
  date: string;
  tempora: Observance[] = [];
  celebration: Observance[] = [];
  commemoration: Observance[] = [];

  constructor(date: string) {
    this.date = date;
  }

  get all(): Observance[] {
    return this.tempora.concat(this.celebration).concat(this.commemoration);
  }

  getTemporaName() {
    return this.tempora[0]?.title;
  }

  serialize() {
    const serialized = {};
    const containers = ["tempora", "celebration", "commemoration"];

    for (const container of containers) {
      serialized[container] = this[container].map((item: Day) =>
        item.serialize(),
      );
    }

    return serialized;
  }
}

class Observance {
  date: string;
  flexibility: string;
  name: string;
  rank: number;
  colors: string[];
  id: string;
  link: string;
  title: string;
  priority: number | null;
  weekday: number | null;

  // """ Build an Observance out of identifier and calendar date
  //
  // :param observance_id: observance's identifier in format
  //                <flexibility>:<identifier>:<rank>
  // :type observance_id: string
  // :param date_: specific date in which the observance is supposed
  //             to be placed. For some Sancti days its rank (class)
  //             depends on which calendar day they occur.
  // :type date_: `date ` object
  // """
  constructor(observanceId: string, date_: string) {
    const [flexibility, name, rank, color] = observanceId.split(":");
    this.date = date_;
    this.flexibility = flexibility;
    this.name = name;
    this.rank = this.calcRank(observanceId, Number.parseInt(rank, 10));
    this.colors = Array.from(color);
    this.id = `${this.flexibility}:${this.name}:${this.rank}:${color}`;
    this.link = `missal/${this.getTempora(this.name)}/${this.name}`;
    this.title = TITLES[observanceId];

    // Determine the weekday attribute based on the type of observance
    if (
      flexibility === TYPE_TEMPORA &&
      ![
        COMMUNE_C_10A,
        COMMUNE_C_10B,
        COMMUNE_C_10C,
        COMMUNE_C_10PASC,
        COMMUNE_C_10T,
      ].includes(observanceId)
    ) {
      if (name.replace(/^.*-(\d+).*$/, "$1") === "10-Dur") {
        this.weekday = 0;
      }
      this.weekday = Number.parseInt(name.replace(/^.*-(\d+).*$/, "$1"), 10);
    } else {
      this.weekday = getDay(this.date);
    }

    this.priority = this.calcPriority();
  }

  private getTempora(name: string): string {
    if (this.flexibility === "tempora") {
      if (name.includes("nat")) {
        return "natal";
      }
      if (name.includes("adv")) {
        return "advento";
      }
      if (name.includes("pasc")) {
        return "pascoa";
      }
      if (name.includes("epi")) {
        return "epifania";
      }
      if (name.includes("quadp")) {
        return "pre-quaresma";
      }
      if (name.includes("quad")) {
        return "quaresma";
      }
      if (name.includes("pent")) {
        return "pentecostes";
      }
    }

    if (this.flexibility === "santos") {
      console.log(`Matched Santos: ${name}`);
      return "santos";
    }

    if (this.flexibility === "commune") {
      return "comum";
    }

    if (this.flexibility === "votivas") {
      return "votivas";
    }

    console.log(`Checking name: ${name}, flexibility: ${this.flexibility}`);
    throw Error(`No directory for ${name}, flexibility , ${this.flexibility}`);
  }

  private calcRank(observanceId: string, originalRank: number): number {
    for (const caseItem of TEMPORA_RANK_MAP) {
      if (
        getMonth(this.date) === caseItem.month &&
        getDate(this.date) === caseItem.day &&
        new RegExp(caseItem.pattern).test(observanceId)
      ) {
        return caseItem.rank;
      }
    }
    return originalRank;
  }

  private calcPriority() {
    for (let priority = 0; priority < TABLE_OF_PRECEDENCE.length; priority++) {
      if (this.id.match(TABLE_OF_PRECEDENCE[priority])) {
        return priority;
      }
    }
  }

  serialize() {
    return {
      id: this.id,
      link: this.link,
      rank: this.rank,
      title: this.title,
      colors: this.colors,
    };
  }
}

export { Day, Observance };
