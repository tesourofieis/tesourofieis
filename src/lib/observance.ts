import { getDate, getDay, getMonth } from "date-fns";
import {
  COMMUNE_C_10A,
  COMMUNE_C_10B,
  COMMUNE_C_10C,
  COMMUNE_C_10PASC,
  COMMUNE_C_10T,
  LANGUAGE,
  PATTERN_POST_EPIPHANY_SUNDAY,
  TABLE_OF_PRECEDENCE,
  TEMPORA_PENT23_0,
  TEMPORA_RANK_MAP,
  TYPE_TEMPORA,
  WEEKDAY_MAPPING,
  getTranslation,
} from "./constants.ts";
import { Proper, ProperConfig } from "./proper.ts";
import { ProperParser } from "./properParser.ts";

export class Observance {
  date: string;
  flexibility: string;
  name: string;
  rank: number;
  colors: string[];
  id: string;
  title: string;
  priority: number | null;
  weekday: number | null;

  // """ Build an Observance out of identifier and calendar date
  //
  // :param observance_id: observance's identifier in format
  //                <flexibility>:<identifier>:<rank>
  // :type observance_id: string
  // :param date_: specific date in which the observance is supposed
  //             to be placed. For some Sancti days its rank (class)
  //             depends on which calendar day they occur.
  // :type date_: `date ` object
  // """
  constructor(observanceId: string, date_: string) {
    const [flexibility, name, rank, color] = observanceId.split(":");
    this.date = date_;
    this.flexibility = flexibility;
    this.name = name;
    this.rank = this.calcRank(observanceId, parseInt(rank, 10));
    this.colors = Array.from(color);
    this.id = `${this.flexibility}:${this.name}:${this.rank}:${color}`;
    this.title = getTranslation(LANGUAGE).TITLES[observanceId];

    // Determine the weekday attribute based on the type of observance
    if (
      flexibility === TYPE_TEMPORA &&
      ![
        COMMUNE_C_10A,
        COMMUNE_C_10B,
        COMMUNE_C_10C,
        COMMUNE_C_10PASC,
        COMMUNE_C_10T,
      ].includes(observanceId)
    ) {
      this.weekday =
        WEEKDAY_MAPPING[parseInt(name.replace(/^.*-(\d+).*$/, "$1"), 10)];
    } else {
      this.weekday = getDay(this.date);
    }

    this.priority = this.calcPriority();
  }

  getProper(config?: ProperConfig) {
    const date = this.date;
    const proper = new ProperParser(this.id, config).parse();
    if (this.id.match(PATTERN_POST_EPIPHANY_SUNDAY) && getMonth(date) >= 9) {
      this.adjustSundayShiftedFromPostEpiphany(proper);
    }
    return proper;
  }

  hasProper(): boolean {
    return new ProperParser(this.id).properExists();
  }

  private calcRank(observanceId: string, originalRank: number): number {
    for (const caseItem of TEMPORA_RANK_MAP) {
      if (
        getMonth(this.date) === caseItem.month &&
        getDate(this.date) === caseItem.day &&
        new RegExp(caseItem.pattern).test(observanceId)
      ) {
        return caseItem.rank;
      }
    }
    return originalRank;
  }

  private calcPriority() {
    for (let priority = 0; priority < TABLE_OF_PRECEDENCE.length; priority++) {
      if (this.id.match(TABLE_OF_PRECEDENCE[priority])) {
        return priority;
      }
    }
  }

  private adjustSundayShiftedFromPostEpiphany(propers: Proper[]): void {
    const properSunday23PostPentecost = new ProperParser(
      TEMPORA_PENT23_0,
    ).parse();
    for (let i = 0; i < propers.length; i++) {
      const proper = propers[i];
      for (const section of [
        "INTROIT",
        "GRADUALE",
        "OFFERTORIUM",
        "COMMUNIO",
      ]) {
        proper.setSection(
          section,
          properSunday23PostPentecost[i][0]?.getSection(section),
        );
      }
    }
  }

  serialize() {
    return {
      id: this.id,
      rank: this.rank,
      title: this.title,
      colors: this.colors,
    };
  }
}
