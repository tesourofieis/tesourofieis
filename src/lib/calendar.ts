import { Day, Observance } from "./day.ts";
import { yyyyMMDD } from "./utils.ts";

import {
  ADVENT,
  EMBER_DAYS_SEPTEMBER,
  FROM_PRE_LENT_TO_POST_PENTECOST,
  LOCAL,
  NATIVITY_OCTAVE_SUNDAY,
  OUTRO,
  POST_EPIPHANY,
  SANCTI,
  SANCTI_10_DUr,
  TEMPORA_NAT2_0,
  WEEK_24_AFTER_PENTECOST,
} from "./constants.ts";
import { rules } from "./rules.ts";

import { UTCDate } from "@date-fns/utc";
import {
  addDays,
  format,
  getDate,
  getDay,
  getMonth,
  getYear,
  isSameDay,
  isSunday,
  nextSunday,
  nextWednesday,
  previousSaturday,
  previousSunday,
  subDays,
} from "date-fns";

class Calendar {
  private container: Map<string, Day>;
  public year: number;

  constructor(year: number) {
    this.year = year;
    this.container = new Map();
    this.buildEmptyCalendar(year);
    this.create();
  }

  private buildEmptyCalendar(year: number): void {
    let date = new UTCDate(year, 0, 1);

    while (date.getFullYear() === year) {
      this.container.set(yyyyMMDD(date), new Day(yyyyMMDD(date)));
      date = addDays(date, 1);
    }
  }

  private create() {
    this.fillInTemporaDays();
    this.fillInSanctiDays();
    this.resolveConcurrency();
  }

  private fillInTemporaDays() {
    // """
    // Days depending on variable date, such as Easter or Advent
    // """
    // # Inserting blocks
    this.insertBlock(this.calcHolyFamily(this.year), POST_EPIPHANY);
    this.insertBlock(
      this.calcSeptuagesima(this.year),
      FROM_PRE_LENT_TO_POST_PENTECOST,
    );
    this.insertBlock(
      this.calcSaturdayBefore24SundayAfterPentecost(this.year),
      POST_EPIPHANY,
      true,
      false,
      this.calcFirstAdventSunday(this.year),
    );
    this.insertBlock(
      this.calc24SundayAfterPentecost(this.year),
      WEEK_24_AFTER_PENTECOST,
    );
    this.insertBlock(
      this.calcFirstAdventSunday(this.year),
      ADVENT,
      false,
      false,
      new UTCDate(this.year, 11, 23),
    );
    this.insertBlock(
      this.calcEmberWednesdaySeptember(this.year),
      EMBER_DAYS_SEPTEMBER,
    );

    // # Inserting single days
    const holyNameDate = this.calcHolyName(this.year);

    this.container.get(yyyyMMDD(holyNameDate)).celebration = [
      new Observance(TEMPORA_NAT2_0, yyyyMMDD(holyNameDate)),
    ];

    const christKingDate = this.calcChristKing(this.year);
    if (this.container.get(yyyyMMDD(christKingDate))) {
      this.container.get(yyyyMMDD(christKingDate)).celebration = [
        new Observance(SANCTI_10_DUr, yyyyMMDD(christKingDate)),
      ];
    }

    const christmasOctaveSunday = this.calcSundayChristmasOctave(this.year);
    if (christmasOctaveSunday) {
      this.insertBlock(christmasOctaveSunday, NATIVITY_OCTAVE_SUNDAY);
    }
  }

  private insertBlock(
    date: Date,
    block: string[],
    reverse = false,
    overwrite = true,
    stopDate?: Date,
  ) {
    if (reverse) {
      // TODO: use toReversed in order not to mutate the original
      // or slice
      block = block.slice().reverse();
    }

    for (const [ii, observanceIds] of block.entries()) {
      const currentDate = addDays(new UTCDate(date), reverse ? -ii : ii);

      if (!observanceIds) {
        continue;
      }

      const dateKey = yyyyMMDD(currentDate);

      if (this.container.get(dateKey)?.celebration.length && !overwrite) {
        break;
      }

      if (isSameDay(stopDate, subDays(currentDate, 1))) {
        break;
      }

      if (this.container.get(dateKey)) {
        this.container.get(dateKey).tempora = [
          new Observance(observanceIds, dateKey),
        ];
        this.container.get(dateKey).celebration = [
          ...this.container.get(dateKey).tempora,
        ];
      }
    }
  }

  private fillInSanctiDays() {
    for (const [date] of this.container.entries()) {
      const dateId = format(date, "MM-dd");
      const days = SANCTI.filter((ii) => ii.startsWith(`santos:${dateId}`)).map(
        (ii) => new Observance(ii, date),
      );

      this.container.get(date)?.celebration.push(...days);

      const local = LOCAL.filter((ii) => ii.startsWith(`santos:${dateId}`)).map(
        (ii) => new Observance(ii, date),
      );

      if (local.length) {
        this.container.get(date)?.local.push(...local);
      }

      const outro = OUTRO.filter((ii) => ii.startsWith(`santos:${dateId}`)).map(
        (ii) => new Observance(ii, date),
      );

      if (outro.length) {
        this.container.get(date)?.outro.push(...outro);
      }
    }
  }

  // Apply `kalendar.rules.*` to the initially instantiated Calendar to fix the situations
  // where more than one Observance falls in the same day.
  private resolveConcurrency() {
    const shiftedAll: { [date: string]: Observance[] } = {};

    for (const [date] of this.container.entries()) {
      const [celebration, commemoration, shifted] = this.applyRules(
        date,
        shiftedAll[date] || [],
      );

      this.container.get(date).celebration = celebration as Observance[];
      this.container.get(date).commemoration = commemoration as Observance[];

      if (shifted) {
        // @ts-ignore
        for (const [k, v] of shifted) {
          if (!shiftedAll[k]) {
            shiftedAll[k] = [];
          }
          shiftedAll[k].push(...(Array.isArray(v) ? v : [v]));
        }
      }
    }
  }

  private applyRules(date: string, shifted: Observance[]) {
    for (const rule of rules) {
      const results = rule(
        this,
        date,
        this.container.get(date)?.tempora,
        this.container.get(date).celebration.concat(shifted),
      );

      if (results) {
        return results;
      }
    }

    return [this.container.get(date).celebration, [], []];
  }

  private calcEasterSunday(year: number): Date {
    const a = year % 19;
    const b = Math.floor(year / 100);
    const c = year % 100;
    const d = Math.floor(b / 4);
    const e = b % 4;
    const f = Math.floor((b + 8) / 25);
    const g = Math.floor((b - f + 1) / 3);
    const h = (19 * a + b - d - g + 15) % 30;
    const i = Math.floor(c / 4);
    const k = c % 4;
    const l = (32 + 2 * e + 2 * i - h - k) % 7;
    const m = Math.floor((a + 11 * h + 22 * l) / 451);
    const month = Math.floor((h + l - 7 * m + 114) / 31);
    const day = ((h + l - 7 * m + 114) % 31) + 1;
    const easter = new UTCDate(year, month - 1, day);

    return easter;
  }

  private calcHolyFamily(year: number): Date {
    // Feast of the Holy Family - First Sunday after Epiphany (06 January).
    const d = new UTCDate(year, 0, 6); // Months are zero-indexed in JavaScript

    return nextSunday(d);
  }

  private calcSeptuagesima(year: number): Date {
    // """ Septuagesima Sunday.
    //
    // Beginning of the pre-Lenten season (Shrovetide).
    // It's 63 days before Ressurection, ninth Sunday before Easter, the third before Ash Wednesday.
    // First day of the Ressurection Sunday - related block.
    // """
    const easter = this.calcEasterSunday(year);
    return subDays(easter, 63);
  }

  private calcFirstAdventSunday(year: number): Date {
    // """
    // First Sunday of Advent - November 27 if it's Sunday, otherwise closest Sunday.
    // """
    const christmasDay = new UTCDate(year, 11, 25);
    const weekday = getDay(christmasDay);

    switch (weekday) {
      case 0: // Sunday
        return new UTCDate(getYear(christmasDay), 10, 27);
      case 1: // Monday
        return new UTCDate(getYear(christmasDay), 11, 3);
      case 2: // Tuesday
        return new UTCDate(getYear(christmasDay), 11, 2);
      case 3: // Wednesday
        return new UTCDate(getYear(christmasDay), 11, 1);
      case 4: // Thursday
        return new UTCDate(getYear(christmasDay), 10, 30);
      case 5: // Friday
        return new UTCDate(getYear(christmasDay), 10, 29);
      default: // Saturday
        return new UTCDate(getYear(christmasDay), 10, 28);
    }
  }

  private calc24SundayAfterPentecost(year: number): Date {
    // """ 24th Sunday after Pentecost.
    //
    // Last Sunday before First Sunday of Advent.
    // It will be always TEMPORA_PENT24_0, which will be placed either:
    // * instead of TEMPORA_PENT23_0
    //   if the number of TEMPORA_PENT*_0 Sundays in given year == 23)
    // * directly after a week starting with TEMPORA_PENT23_0
    //   if the number of TEMPORA_PENT*_0 Sundays in given year == 24)
    // * directly after a week starting with TEMPORA_EPI6_0 (moved from post-epiphania period)
    //   if the number of TEMPORA_PENT*_0 Sundays in given year > 24)
    // """
    return previousSunday(new UTCDate(this.calcFirstAdventSunday(year)));
  }

  private calcSaturdayBefore24SundayAfterPentecost(year: number): Date {
    // """ Last Saturday before 24th Sunday after Pentecost.
    //
    // This is the end of potentially "empty" period that might appear
    // between 23rd and 24th Sunday after Pentecost if Easter is early.
    // In such case one or more Sundays after Epiphany (TEMPORA_EPI*_0) are moved here to "fill the gap"

    return previousSaturday(new UTCDate(this.calc24SundayAfterPentecost(year)));
  }

  private calcEmberWednesdaySeptember(year: number): Date {
    // """ Wednesday of the Ember Days of September.
    //
    // Ember Wednesday in September is a Wednesday after third Sunday
    // of September according to John XXIII's motu proprio
    // "Rubricarum instructum" of June 25 1960.
    // """
    let d = new UTCDate(year, 8, 1);
    while (getMonth(d) === 8) {
      if (isSunday(d) && getDate(d) >= 15 && getDate(d) <= 21) {
        break;
      }
      d = addDays(d, 1);
    }
    return nextWednesday(d);
  }

  private calcHolyName(year: number) {
    // """ The Feast of the Holy Name of Jesus.
    //
    // Kept on the First Sunday of the year; but if this Sunday falls on
    // 1st, 6th or 7th January, the feast is kept on 2nd January.
    // """
    let d = new UTCDate(year, 0, 1);

    while (getDate(d) <= 7) {
      if (isSunday(d)) {
        if (getDate(d) === 1 || getDate(d) === 6 || getDate(d) === 7) {
          return new UTCDate(year, 0, 2);
        }

        return d;
      }

      d = addDays(d, 1);
    }

    console.error("Unexpected condition: Holy Name date not found.");
  }

  private calcChristKing(year: number): Date {
    // The Feast of Christ the King, last Sunday of October.
    const d = new UTCDate(year, 10, 1);

    return previousSunday(d);
  }

  private calcSundayChristmasOctave(year: number): Date | null {
    // """
    // Sunday within the Octave of Christmas, falls between Dec 26 and Dec 31
    // """
    let d = new UTCDate(year, 11, 27); // December 27
    while (d.getFullYear() === year) {
      if (isSunday(d)) {
        return d;
      }
      d = addDays(d, 1);
    }
    return null;
  }

  get(date: string) {
    return this.container.get(date);
  }

  findDay(observanceId: string) {
    for (const [date_, day] of this.container) {
      if (day.all.some((observance) => observance.id === observanceId)) {
        return [date_, day];
      }
    }
  }

  private items() {
    return this.container.entries();
  }

  serialize(): { [date: string]: Day } {
    const serialized = {};
    for (const [date_, day] of this.items()) {
      serialized[date_] = day.serialize();
    }
    return serialized;
  }
}

export { Calendar };
