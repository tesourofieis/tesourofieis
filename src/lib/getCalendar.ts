import { Calendar } from "./src/calendar.ts";

function getCalendar(year: number) {
  const calendar = new Calendar(year);

  return calendar;
}

function getCalendarDay(year: number, date: string) {
  const calendar = new Calendar(year).get(date);

  return calendar;
}

export { getCalendar, getCalendarDay };
