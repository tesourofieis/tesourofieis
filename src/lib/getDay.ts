import { getYear } from "date-fns";
import { Calendar } from "./calendar.ts";
import { format_propers } from "./utils.ts";

function getDay(date: string) {
  const calendar = new Calendar(getYear(date));

  const day = calendar.get(date);

  const proper = day?.getProper(calendar);

  const format = format_propers(proper);

  return format;
}

export { getDay };
