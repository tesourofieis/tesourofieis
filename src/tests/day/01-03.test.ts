import { expect, test } from "vitest";
import { getDay } from "../../lib/getDay";

test("2024-01-03", () => {
  const proper = getDay("2024-01-03");

  expect(proper).toMatchSnapshot();
});
