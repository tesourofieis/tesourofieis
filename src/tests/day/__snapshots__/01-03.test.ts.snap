// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`2024-01-03 1`] = `
[
  {
    "info": {
      "colors": [
        "w",
      ],
      "date": "2024-01-03",
      "day": Day {
        "celebration": [],
        "commemoration": [],
        "date": "2024-01-03",
        "tempora": [],
      },
      "id": "sancti:01-01:1:w",
      "rank": 4,
      "tempora": undefined,
      "title": "Circuncisão do Senhor e Oitava do Natal",
    },
    "sections": [
      {
        "body": {
          "latin": [
            "Isa 9:6",
            "Puer natus est nobis, et fílius datus est nobis: cujus impérium super húmerum ejus: et vocábitur nomen ejus magni consílii Angelus.",
            "Ps 97:1",
            "Cantáte Dómino cánticum novum: quia mirabília fecit.",
            "Glória Patri…",
            "Puer natus est nobis, et fílius datus est nobis: cujus impérium super húmerum ejus: et vocábitur nomen ejus magni consílii Angelus.",
            "",
          ],
          "vernacular": [
            "Is 9:6",
            "Nasceu para nós um Menino e foi-nos dado um Filho que traz sobre os ombros o manto da realeza, o qual será chamado «Anjo do admirável conselho».",
            "Sl 97:1",
            "Cantai ao Senhor um hino novo, pois Ele operou maravilhas.",
            "Glória ao Pai.",
            "Nasceu para nós um Menino e foi-nos dado um Filho que traz sobre os ombros o manto da realeza, o qual será chamado «Anjo do admirável conselho».",
            "",
          ],
        },
        "id": "Introitus",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Deus, qui salútis ætérnæ, beátæ Maríæ virginitáte fecúnda, humáno géneri præmia præstitísti: tríbue, quǽsumus; ut ipsam pro nobis intercédere sentiámus, per quam merúimus auctórem vitæ suscípere, Dóminum nostrum Jesum Christum, Fílium tuum:",
            "Qui tecum…",
            "",
          ],
          "vernacular": [
            "Ó Deus, que pela virgindade fecunda da Bem-aventurada Virgem Maria concedestes ao género humano o prémio da salvação eterna, permiti, Vos imploramos, que gozemos a intercessão daquela pela qual fomos julgados dignos de receber o autor da vida, nosso Senhor Jesus Cristo, vosso Filho:",
            "Que convosco vive...",
            "",
          ],
        },
        "id": "Oratio",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Léctio Epístolæ beati Pauli Apóstoli ad Titum",
            "Tit 2:11-15",
            "Caríssime: Appáruit grátia Dei Salvatóris nostri ómnibus homínibus, erúdiens nos, ut, abnegántes impietátem et sæculária desidéria, sóbrie et juste et pie vivámus in hoc sǽculo, exspectántes beátam spem et advéntum glóriæ magni Dei et Salvatóris nostri Jesu Christi: qui dedit semetípsum pro nobis: ut nos redímeret ab omni iniquitáte, et mundáret sibi pópulum acceptábilem, sectatórem bonórum óperum. Hæc lóquere et exhortáre: in Christo Jesu, Dómino nostro.",
            "",
          ],
          "vernacular": [
            "Lição da Ep.ª do B. Ap.º Paulo a Tito.",
            "Tt 2:11-15",
            "Caríssimo: A graça de Deus, nosso Salvador, manifestou-se a todos os homens, ensinando-nos, a fim de que, repudiando a impiedade e os apetites terrenos, vivamos neste mundo com temperança, justiça e piedade, pensando na esperança, na bem-aventurança eterna e na vinda da glória do nosso grande Deus e Salvador, Jesus Cristo: que se ofereceu espontaneamente por nós, para nos resgatar de toda a iniquidade e tornar-nos numa raça purificada, escolhida e zelosa em suas boas obras. Ensina e prega estas cousas, em Jesus Cristo, nosso Senhor!",
            "",
          ],
        },
        "id": "Lectio",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Ps 97:3; 97:2",
            "Vidérunt omnes fines terræ salutare Dei nostri: jubiláte Deo, omnis terra.",
            "℣. Notum fecit Dominus salutare suum: ante conspéctum géntium revelávit justitiam suam. Allelúja, allelúja.",
            "Heb 1:1-2",
            "Multifárie olim Deus loquens pátribus in Prophétis, novíssime diébus istis locútus est nobis in Fílio. Allelúja.",
            "",
          ],
          "vernacular": [
            "Sl 97:3; 97:2",
            "Toda a terra viu o Salvador que o nosso Deus enviou: aclamai, pois, o Senhor, ó povos de toda a terra!",
            "℣. O Senhor manifestou o Salvador que havia prometido: e manifestou a sua justiça aos olhos dos povos. Aleluia, aleluia.",
            "Heb 1:1-2",
            "Deus, que outrora falou de diversas maneiras pelos Profetas, dignou-se falar-nos nos últimos tempos pelo seu Filho. Aleluia.",
            "",
          ],
        },
        "id": "Graduale",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Sequéntia ☩ sancti Evangélii secúndum Lucam",
            "Luc 2:21",
            "In illo témpore: Postquam consummáti sunt dies octo, ut circumciderétur Puer: vocátum est nomen ejus Jesus, quod vocátum est ab Angelo, priúsquam in útero conciperétur.",
            "",
          ],
          "vernacular": [
            "Continuação ☩ do santo Evangelho segundo S. Lucas.",
            "Lc 2:21",
            "Naquele tempo, passados que foram oito dias depois dos quais o Menino devia ser circuncidado, foi-Lhe dado o nome de Jesus, que foi aquele que o Anjo Lhe havia dado, antes de ser concebido no seio de sua Mãe.",
            "",
          ],
        },
        "id": "Evangelium",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Ps 88:12; 88:15",
            "Tui sunt coeli et tua est terra: orbem terrárum et plenitúdinem ejus tu fundásti: justítia et judícium præparátio sedis tuæ.",
            "",
          ],
          "vernacular": [
            "Sl 88:12; 88:15",
            "A Vós, Senhor, pertencem os céus e a terra, pois criastes o universo e tudo o que ele encerra. A justiça e a equidade são a base do vosso trono.",
            "",
          ],
        },
        "id": "Offertorium",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Munéribus nostris, quǽsumus, Dómine, precibúsque suscéptis: et coeléstibus nos munda mystériis, et cleménter exáudi.",
            "Per Dominum…",
            "",
          ],
          "vernacular": [
            "Havendo Vós, Senhor, recebido benignamente os nossos dons e orações, dignai-Vos ainda, Vos suplicamos, purificar-nos pela virtude dos vossos celestiais mystérios e ouvir-nos misericordiosamente.",
            "Por nosso Senhor...",
            "",
          ],
        },
        "id": "Secreta",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Communis",
            "Vere dignum et justum est, æquum et salutáre, nos tibi semper et ubíque grátias ágere: Dómine sancte, Pater omnípotens, ætérne Deus: per Christum, Dóminum nostrum. Per quem majestátem tuam laudant Angeli, adórant Dominatiónes, tremunt Potestátes. Cæli cælorúmque Virtútes ac beáta Séraphim sócia exsultatióne concélebrant. Cum quibus et nostras voces ut admítti jubeas, deprecámur, súpplici confessione dicéntes:",
            "",
          ],
          "vernacular": [
            "Comum",
            "É verdadeiramente digno e justo, racional e salutar render-Vos graças em todos os lugares e sempre, ó Senhor santo, Pai omnipotente, eterno Deus, por nosso Senhor Jesus Cristo. Pelo qual os Anjos louvam a vossa majestade, as Dominações a adoram e as Potestades se prostram reverentes; e os Céus, as Virtudes dos céus e os bem-aventurados Serafins a celebram em comum alegria e louvor. Dignai-Vos permitir, Vos imploramos, que as nossas vozes suplicantes se unam às deles, dizendo:",
            "",
          ],
        },
        "id": "Prefatio",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Ps 97:3",
            "Vidérunt omnes fines terræ salutáre Dei nostri.",
            "",
          ],
          "vernacular": [
            "Sl 97:3",
            "Toda a terra viu o Salvador que o nosso Deus enviou.",
            "",
          ],
        },
        "id": "Communio",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Hæc nos commúnio, Dómine, purget a crímine: et, intercedénte beáta Vírgine Dei Genetríce María, cæléstis remédii fáciat esse consórtes.",
            "Per eúndem…",
            "",
          ],
          "vernacular": [
            "Que esta comunhão, Senhor, nos purifique de nossos crimes e que por intercessão da B. Virgem Maria, Mãe de Deus, nos torne participantes do remédio celestial.",
            "Pelo mesmo nosso Senhor...",
            "",
          ],
        },
        "id": "Postcommunio",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Sequéntia ☩ sancti Evangélii secúndum Matthǽum",
            "Matt 2:1-12",
            "Cum natus esset Jesus in Béthlehem Juda in diébus Heródis regis, ecce, Magi ab Oriénte venerunt Jerosólymam, dicéntes: Ubi est, qui natus est rex Judæórum? Vidimus enim stellam ejus in Oriénte, et vénimus adoráre eum. Audiens autem Heródes rex, turbatus est, et omnis Jerosólyma cum illo. Et cóngregans omnes principes sacerdotum et scribas pópuli, sciscitabátur ab eis, ubi Christus nasceretur. At illi dixérunt ei: In Béthlehem Judæ: sic enim scriptum est per Prophétam: Et tu, Béthlehem terra Juda, nequaquam mínima es in princípibus Juda; ex te enim éxiet dux, qui regat pópulum meum Israël. Tunc Heródes, clam vocátis Magis, diligénter dídicit ab eis tempus stellæ, quæ appáruit eis: et mittens illos in Béthlehem, dixit: Ite, et interrogáte diligénter de púero: et cum invenéritis, renuntiáte mihi, ut et ego véniens adórem eum. Qui cum audíssent regem, abiérunt. Et ecce, stella, quam víderant in Oriénte, antecedébat eos, usque dum véniens staret supra, ubi erat Puer. Vidéntes autem stellam, gavísi sunt gáudio magno valde. Et intrántes domum, invenérunt Púerum cum María Matre ejus, (hic genuflectitur) et procidéntes adoravérunt eum. Et, apértis thesáuris suis, obtulérunt ei múnera, aurum, thus et myrrham. Et respónso accépto in somnis, ne redírent ad Heródem, per aliam viam revérsi sunt in regiónem suam.",
            "",
            "",
          ],
          "vernacular": [
            "Continuação ☩ do santo Evangelho segundo S. Mateus.",
            "Mt. 2:19-23",
            "Havendo Jesus nascido em Belém, de Judá, no tempo do rei Herodes, eis que vieram a Jerusalém os Magos do Oriente, dizendo: «Onde está o Rei dos Judeus, que acaba de nascer? Pois vimos a sua estrela no Oriente e viemos adorá-l’O». Logo que o rei Herodes ouviu esta notícia, ficou perturbado, assim como toda a gente de Jerusalém, convocando logo todos os príncipes dos sacerdotes e os escribas do povo, para saber deles onde deveria nascer Cristo. Responderam-lhe eles: «Em Belém, de Judá, pois está escrito pelo Profeta: «E tu, Belém, terra de Judá, não serás certamente a menos importante entre as terras principais de Judá, pois em ti nascerá o Rei, que governará o meu povo de Israel». Então Herodes mandou chamar em segredo os Magos, informando-se com eles diligentemente acerca do tempo em que a estrela havia aparecido. E enviando-os a Belém, disse-lhes: «Ide, procurai diligentemente o Menino, e, logo que O houverdes achado, avisai-me, para que eu vá, também, adorá-l’O». Os Magos, tendo ouvido estas palavras, partiram. Ora, a estrela, que tinham visto no Oriente ia adiante deles, até que, chegando ao lugar onde estava o Menino, parou. Quando os Magos viram a estrela, alegraram-se muito. Entrando, então, na casa, encontraram o Menino com Maria, sua mãe; e, de joelhos, O adoraram. (Todos devem ajoelhar) E, tendo aberto os seus tesouros, ofereceram-Lhe presentes de ouro, incenso e mirra. Depois, havendo tido aviso em sonhos de que não deveriam voltar a encontrar Herodes, retiraram-se por outro caminho para o seu país.",
          ],
        },
        "id": "Ultima Evangelium",
        "subSections": [],
      },
    ],
  },
]
`;
